File: dumpBuilder.py
import glob
import yaml
import pprint
import os
import zipfile

BASE_DIRECTORIES = [
    "components",
    "missiles",
    "stations",
    "resources",
]

filenames = []
for directory in BASE_DIRECTORIES:
    filenames.extend(
        glob.glob(f"{directory}/**/*.fw", recursive=True)
    )


result = {
    "resources": {},
    "components": {},
    "combat_stations": {},
    "missiles": {},
}

for filename in filenames:
    print(f"{filename}...")
    with open(filename) as f:
        current = yaml.load(f, Loader=yaml.SafeLoader)
        if("resources" in current):
            result["resources"].update(
                current["resources"]
            )
        if("component" in current):
            result["components"].update(
                current["component"]
            )
        if("components" in current):
            result["components"].update(
                current["components"]
            )
        if("missile" in current):
            result["missiles"].update(
                current["missile"]
            )
        if("missiles" in current):
            result["missiles"].update(
                current["missiles"]
            )
        if("missiles" in current):
            result["missiles"].update(
                current["missiles"]
            )
        if("combat_station" in current):
            result["combat_stations"].update(
                current["combat_station"]
            )
        if("combat_stations" in current):
            result["combat_stations"].update(
                current["combat_stations"]
            )

with open("junctspace.yaml", "w") as f:
    yaml.dump(result, f, Dumper=yaml.SafeDumper, sort_keys=False)

with open("junctspace.txt", "w") as f:
    yaml.dump(result, f, Dumper=yaml.SafeDumper, sort_keys=False)

with open("junctspace.fw", "w") as f:
    yaml.dump(result, f, Dumper=yaml.SafeDumper, sort_keys=False)

with open("junctspace_dump.yaml", "w") as f:
    yaml.dump(result, f, Dumper=yaml.SafeDumper, sort_keys=False)

with open("junctspace_dump.txt", "w") as f:
    yaml.dump(result, f, Dumper=yaml.SafeDumper, sort_keys=False)

with open("junctspace_dump.fw", "w") as f:
    yaml.dump(result, f, Dumper=yaml.SafeDumper, sort_keys=False)

path = ""
zipf = zipfile.ZipFile('junctspace.zip'.format(os.path.join(path, directory)), 'w', zipfile.ZIP_DEFLATED)
for filename in filenames:
    zipf.write(os.path.abspath(filename), arcname=filename)
zipf.close()

File: dumpBuilder.js
const glob = require("glob");
const yaml = require("js-yaml");
const fs = require("fs");
const path = require("path");
const AdmZip = require("adm-zip");

// Gamedata master directory
const DATA_DIR = "gamedata/";

// Vanilla Gamedata Processing
const BASEDATA_DIR = DATA_DIR + "fleetwright/";

const BASE_DIRECTORIES = [
  "components",
  "missiles",
  "stations",
  "resources",
];

let filenames = [];
for (const directory of BASE_DIRECTORIES) {
  filenames = filenames.concat(
    glob.sync(`${BASEDATA_DIR + directory}/**/*.fw`, { recursive: true })
  );
}

const result = {
  resources: {},
  components: {},
  combat_stations: {},
  missiles: {},
};

for (const filename of filenames) {
  console.log(`${filename}...`);
  const current = yaml.safeLoad(fs.readFileSync(filename, "utf8"));
  if ("resources" in current) {
    Object.assign(result.resources, current.resources);
  }
  if ("component" in current) {
    Object.assign(result.components, current.component);
  }
  if ("components" in current) {
    Object.assign(result.components, current.components);
  }
  if ("missile" in current) {
    Object.assign(result.missiles, current.missile);
  }
  if ("missiles" in current) {
    Object.assign(result.missiles, current.missiles);
  }
  if ("combat_station" in current) {
    Object.assign(result.combat_stations, current.combat_station);
  }
  if ("combat_stations" in current) {
    Object.assign(result.combat_stations, current.combat_stations);
  }
}

fs.writeFileSync("junctspace.yaml", yaml.safeDump(result, { sortKeys: false }));

fs.writeFileSync("junctspace.txt", yaml.safeDump(result, { sortKeys: false }));

fs.writeFileSync("junctspace.fw", yaml.safeDump(result, { sortKeys: false }));

fs.writeFileSync(
  "junctspace_dump.yaml",
  yaml.safeDump(result, { sortKeys: false })
);

fs.writeFileSync(
  "junctspace_dump.txt",
  yaml.safeDump(result, { sortKeys: false })
);

fs.writeFileSync(
  "junctspace_dump.fw",
  yaml.safeDump(result, { sortKeys: false })
);

const zip = new AdmZip();
for (const filename of filenames) {
  zip.addLocalFile(path.resolve(filename));
}
zip.writeZip("fleetwright.zip");
